packages/plugin-kafka/__tests__/kafka.spec.ts

```diff
@@ -0,0 +1,197 @@
+/**
+ * @jest-environment node
+ */
+
+// tests long running
+jest.setTimeout(10000)
+
+import { Microfleet } from '@microfleet/core'
+import { KafkaPlugin, KafkaFactory } from '../lib/kafka'
+import { ConsumerStream, ProducerStream, KafkaConsumer, Producer } from 'node-rdkafka'
+import { promisify } from 'util'
+import { pipeline, Readable } from 'stream'
+
+const RdKafkaError = require('node-rdkafka/lib/error')
+
+import sinon, { stub } from 'sinon'
+
+let service: Microfleet | Microfleet & KafkaPlugin
+let producer: ProducerStream
+let consumer: ConsumerStream
+
+// const wait = async (ms: number) => new Promise(resolve => setTimeout(resolve, ms))
+
+beforeEach(async () => {
+  service = new Microfleet({
+    name: 'tester',
+    plugins: ['logger', 'validator', 'kafka'],
+    kafka: {
+      debug: 'consumer,cgrp,topic,fetch',
+      'metadata.broker.list': 'kafka:9092',
+      'group.id': 'test-group',
+    },
+  })
+})
+
+afterEach(async () => {
+  await service.close()
+})
+
+describe('connect', () => {
+  test('should be able to create a producer', async () => {
+    producer = await service.kafka.createProducerStream({ objectMode: false, topic: 'testBoo' })
+    expect(producer).toBeDefined()
+  })
+
+  test('should be able to create a consumer', async () => {
+    consumer = await service.kafka.createConsumerStream({ topics: ['test'] })
+    expect(consumer).toBeDefined()
+  })
+
+  test('tracks streams ', async () => {
+    const kafka: KafkaFactory = service.kafka
+
+    const streamToClose = await kafka.createProducerStream(
+      { objectMode: false, topic: 'otherTopicBar' },
+      { 'group.id': 'track-group' },
+      {}
+    )
+    const streamToCloseToo = await kafka.createConsumerStream(
+      { topics: 'otherTopicBar' },
+      { 'group.id': 'track-group' },
+      { 'auto.offset.reset': 'earliest' }
+    )
+
+    expect(kafka.getStreams().size).toEqual(2)
+
+    await promisify(streamToClose.close.bind(streamToClose))()
+    expect(kafka.getStreams().size).toEqual(1)
+
+    await promisify(streamToCloseToo.close.bind(streamToCloseToo))()
+    expect(kafka.getStreams().size).toEqual(0)
+  })
+
+  test('closes streams on service shutdown', async () => {
+    const kafka: KafkaFactory = service.kafka
+
+    await kafka.createProducerStream(
+      { objectMode: false, topic: 'otherTopicBaz' },
+      { 'group.id': 'track-close-group' }
+    )
+
+    await kafka.createConsumerStream(
+      { topics: 'otherTopicBaz' },
+      { 'group.id': 'track-close-group' },
+      { 'auto.offset.reset': 'earliest' }
+    )
+
+    await service.close()
+
+    expect(kafka.getStreams().size).toEqual(0)
+  })
+
+})
+
+describe('connect error stubbed', () => {
+  let consumerStub: sinon.SinonStub<[any?, (((err: any, data: any) => any) | undefined)?], KafkaConsumer>
+  let producerStub: sinon.SinonStub<[any?, (((err: any, data: any) => any) | undefined)?], Producer>
+
+  beforeAll(() => {
```

Comment:
this could be improved by triggering real errors via toxiproxy or similar

Reply:
fixed